### ISSUE TITLE
Move Crucible.LLVM.MemModel.Value.{anyOf,allOf} to What4

### ISSUE THREAD

**Issue by siddharthist** [GaloisInc/crucible#201]
Leftover from #200: These functions are general-purpose short-circuiting n-ary `and` and `or` operations on What4 `Pred`s. They should be moved into the `ExprBuilder` interface to allow backends to override them if a more efficient implementation of n-ary `and` or `or` is available.

**Comment by robdockins** [comment]
When #176 gets merged, all What4 conjunctions and disjunctions will internally be represented as n-ary operations, so how we construct them will matter less.

I'm interested to know if the constant-detecting phase of `anyOf`/`allOf` actually help in practice over `andAllOf`/`orOneOf`, which just rely on constant-folding of the underlying boolean operation.

**Comment by siddharthist** [comment]
I'm also interested to know for sure, see my comment on #200 for my thinking on why this might be a bit more efficient. Will the refactor come with n-ary constructors via `ExprBuilder`? In that case, these would certainly be obviated by the application of `asConstantPred` to the n-ary conjoined predicate.

**Comment by robdockins** [comment]
Yeah, I've just learned not to necessarily trust my intuition for what is going to be faster.  If your method of checking for constant predicates first is more efficient, we should just roll it into the preexisting `andAllOf`/`orOneOf`.  We could lift these to class methods, I suppose, but I'm not sure there is much value in it.

I wasn't planning to expose a direct n-way operators, as the local rewriting rules are much easier to implement with only two inputs.  Also, the checking for, e.g., absorption instances has to be done one-at-a-time anyway (I think), so I don't think it would buy much.

**Comment by langston-barrett** [comment]
I'm retargeting this issue to simply removing these functions.
